### PR TITLE
fix: skill write atomicity

### PR DESCRIPTION
**Pull request by markus-lassfolk**
## Summary

Closes #

**PR title:** Must follow [Conventional Commits](https://www.conventionalcommits.org/) (e.g. `fix: …`, `chore: merge develop into main`). CI validates the title via the **Conventional Commits** check.

**Large merges:** For branches with many commits (e.g. `develop` → `main`), GitHub may pre-fill the merge/squash description with every commit message. Before completing merge or auto-merge, switch to **Squash and merge** (if allowed) and replace that body with a short summary, or merge locally with `git merge -m "…"`.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal improvement (no behavior change)
- [ ] Documentation update
- [ ] CI / tooling change

## Quality Checklist

Run from repo root (package lives under `extensions/memory-hybrid/`):

- [ ] `cd extensions/memory-hybrid && npx tsc --noEmit` passes with no errors
- [ ] `cd extensions/memory-hybrid && npm run lint` passes with no warnings
- [ ] `cd extensions/memory-hybrid && npm test` passes (all tests green)
- [ ] No secrets, credentials, or API keys committed
- [ ] No `console.log` debug statements left in production code

## Documentation Impact (REQUIRED)

- [ ] Inline code comments (JSDoc) updated for modified functions and types
- [ ] `docs/` or `README.md` updated to reflect architectural or usage changes
- [ ] Cross-referenced functions/services checked for outdated documentation

## Testing

- [ ] Unit tests added / updated
- [ ] Integration tests added / updated
- [ ] Manually verified against a running OpenClaw instance

## Notes for Reviewer

---

> [!NOTE]
> **Medium Risk**
> Changes how skills are persisted to disk by introducing atomic file/directory writes and new completeness markers, which could affect promotion workflows and collision handling if any edge cases are missed. Risk is mitigated by extensive new unit tests around crash leftovers, overwrite refusal, and rollback behavior.
> 
> **Overview**
> **Makes skill generation/writes crash-atomic.** Introduces `utils/atomic-write` with `atomicWriteFile` and `atomicWriteSkillDir` (plus a `.openclaw-skill-complete` marker) to prevent partially-written `SKILL.md`/sidecars from being treated as committed skills.
> 
> Updates crystallization and procedure auto-skill generation to use atomic writes, retry slug allocation on write collisions, and only rollback directories it definitively created (via marker matching). Duplicate/slug-collision detection now ignores atomic temp/backup artifacts unless they have the completion marker, while still treating legacy markerless directories as real skills.
> 
> Adds comprehensive tests for atomic write behavior, collision/rollback edge cases, and adjusts existing tests to create the new completion marker. Also pins `package-lock.json` to Node `>=22.16.0` via `engines`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c37b66527b525a7e54618a2e18d9c1c66b2da920. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by markus-lassfolk** [comment]
@cursoragent review

File: extensions/memory-hybrid/services/procedure-promotion-policy.ts [hunk 1]
```diff
@@ -1109,6 +1109,9 @@ function isDuplicateSkill(
     for (const entry of safeReadDir(dir)) {
       const skillPath = join(dir, entry, "SKILL.md");
       if (!existsSync(skillPath)) continue;
+      // Legacy skill directories may not have the atomic completion marker. If
+      // they contain SKILL.md, treat them as valid duplicates so marker rollout
+      // never overwrites or re-promotes existing skills.
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/procedure-promotion-policy.ts]: Fixed in dad995c — duplicate scanning now ignores incomplete atomic temp/backup skill directories unless they carry the completion marker, while preserving legacy markerless final skill dirs as duplicates.

File: extensions/memory-hybrid/utils/atomic-write.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    // Stamp the completion marker as the final write inside the temp dir.
+    writeFileSync(join(tmpDir, SKILL_COMPLETE_MARKER), new Date().toISOString(), "utf-8");
+
+    // Move existing skill dir out of the way so the rename targets a free path.
+    if (backupDir) renameSync(skillDir, backupDir);
+
+    // Atomic promotion: temp dir → final skill dir.
+    renameSync(tmpDir, skillDir);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/utils/atomic-write.ts]: Fixed in dad995c — atomicWriteSkillDir no longer supports in-place overwrites; it refuses existing target directories instead of moving committed skills aside, preserving crash-atomic create semantics.

File: extensions/memory-hybrid/utils/atomic-write.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  const rand = randomBytes(8).toString("hex");
+  const tmpDir = `${skillDir}.tmp-${process.pid}-${rand}`;
+  const backupDir = existsSync(skillDir) ? join(parent, `.${basename(skillDir)}.bak-${Date.now()}-${rand}`) : null;
+
+  try {
+    // Write every sidecar into the temp directory.
+    for (const [relPath, content] of Object.entries(files)) {
+      const fullPath = join(tmpDir, relPath);
+      mkdirSync(dirname(fullPath), { recursive: true });
+      writeFileSync(fullPath, content, "utf-8");
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/utils/atomic-write.ts:83]: Fixed in dad995c — atomicWriteSkillDir now validates every relPath key as a safe relative path and rejects absolute, parent traversal, empty, null-byte, and completion-marker paths before writing.

**Comment by markus-lassfolk** [comment]
## Council review verdict: BLOCKERS

Reviewed current head `40111b2aac573b570001f816cdbaef7e118baf81` against the atomic skill-write acceptance criteria from #1405 / #1419.

### Blocker: rollback can still delete a concurrently-created legitimate skill directory

`generateAutoSkills` and `generateAutoSkillForProcedure` snapshot `skillDirExistedBefore` before calling `writeDraftSkill`:

- `extensions/memory-hybrid/services/procedure-skill-generator.ts:248-270`
- `extensions/memory-hybrid/services/procedure-skill-generator.ts:389-394`

If `skillDirExistedBefore === false`, but another process creates and commits the same `skillDir` before `atomicWriteSkillDir` reaches its existence check, `atomicWriteSkillDir` correctly throws `Refusing to overwrite existing skill directory` (`utils/atomic-write.ts:72-75`). The catch block then calls:

```ts
rollbackDraftSkill(skillDir, skillDirExistedBefore);
```

and `rollbackDraftSkill` deletes unconditionally for this case because the guard only protects when `skillDirExistedBefore !== false`:

```ts
if (skillDirExistedBefore !== false && isCommittedSkillDir(skillDir)) return;
rmSync(skillDir, { recursive: true, force: true });
```

So a TOCTOU collision can still remove a valid skill directory created by the winning process. That violates the intended crash/concurrency safety of the atomic writer path and is especially risky because this code now refuses overwrites explicitly.

**Actionable fix:** make rollback delete only directories this operation actually created. For example:

- Detect the specific `Refusing to overwrite existing skill directory` error and skip rollback entirely.
- Or have `writeDraftSkill` / `atomicWriteSkillDir` return a committed/created flag and only rollback when the current call successfully created the final dir before a later DB transition failed.
- Add a regression test simulating `existsSync(skillDir) === false` before the try, then `atomicWriteSkillDir` throwing because the target now exists, and assert the existing committed directory is preserved.

### Evidence checked

- CI/status checks on current head are green (type check, lint, tests, coverage, CodeQL, Cursor Bugbot all successful in GitHub rollup).
- The prior issues around unsafe relative paths, in-place overwrite crash atomicity, duplicate artifact detection, and centralized artifact detection appear addressed at current head.
- Local targeted test run was not possible in the cloned review checkout because dependencies were not installed (`vitest: not found`), so I relied on GitHub CI for test execution evidence.

**Comment by markus-lassfolk** [comment]
@cursoragent review

Stewardship request: please review current head `40111b2aac573b570001f816cdbaef7e118baf81` for PR #1441 before Council/deep-verifier advancement.

**Review by markus-lassfolk — COMMENTED** [review]
Reviewed current head `40111b2aac573b570001f816cdbaef7e118baf81` for the Council stage.

I am requesting changes because the current head still has the rollback TOCTOU deletion blocker.

`generateAutoSkills` and `generateAutoSkillForProcedure` snapshot `skillDirExistedBefore = existsSync(skillDir)` before calling `writeDraftSkill`, then call `rollbackDraftSkill(skillDir, skillDirExistedBefore)` on any write/promote failure. At current head, `rollbackDraftSkill` only preserves an existing committed directory when `skillDirExistedBefore !== false`:

```ts
if (skillDirExistedBefore !== false && isCommittedSkillDir(skillDir)) return;
rmSync(skillDir, { recursive: true, force: true });
```

If this process observes `skillDirExistedBefore === false`, but another process creates and commits the same `skillDir` before `atomicWriteSkillDir` reaches its `existsSync(skillDir)` refusal check, `atomicWriteSkillDir` correctly throws `Refusing to overwrite existing skill directory`. The catch block then rolls back with `skillDirExistedBefore === false`, bypasses the committed-dir guard, and deletes the concurrently-created legitimate skill directory.

That means the new atomic writer path can still destroy a valid skill created by a winning concurrent process, so the crash/concurrency safety goal is not met.

Required fix:
- Do not delete on the `Refusing to overwrite existing skill directory` path; or
- Have `writeDraftSkill`/`atomicWriteSkillDir` report whether this invocation actually committed the final directory, and only rollback directories that this invocation created before a later DB transition failed.
- Add a regression test for the TOCTOU case: pre-check returns false, `atomicWriteSkillDir` refuses because the target now exists, and rollback preserves that existing committed directory.

Other evidence checked:
- GitHub checks are green on current head.
- GitHub review threads visible via API are resolved.
- The atomic file/dir helper, path containment validation, artifact detection centralization, legacy markerless behavior, and temp/backup duplicate ignoring otherwise look directionally sound.

**Comment by markus-lassfolk** [comment]
@cursoragent review

Council review requested for current head. Please review implementation correctness, tests, and issue acceptance criteria; post approval or concrete blockers.

Repository: markus-lassfolk/openclaw-hybrid-memory
PR: #1441
Current head: b051f3fe710f3b2a2401dbcac536a77e703d6259

**Comment by markus-lassfolk** [comment]
@cursoragent review

Stewardship advance: PR #1441 ready for Council review at current head `0ac1e4b4771c33152486e0518b40e10b6caa9392`.

**Status at current head:**
- All review threads resolved ✓
- CI green, mergeable, CLEAN ✓
- Carried stale Council evidence from older heads noted (does not block current-head review)
- Labels: `extensions size/XL fixes-needed stage/review`

**PR title:** fix: skill write atomicity
**PR goal:** Prevents partially-written skills from being treated as committed. Introduces `utils/atomic-write.ts` (atomic single-file writes, atomic multi-file skill-dir promotion via temp dir + rename + `.openclaw-skill-complete` marker). Updates crystallization and procedure generation to use atomic writers. Adjusts slug-collision/duplicate detection to ignore incomplete `.tmp-*`/`.bak-*` artifacts while treating legacy markerless skill dirs as occupied.

**Council criteria:** Approve if — atomic write semantics are correct and safe, marker-based duplicate detection is sound, rollback cleanup only removes dirs written by this run, tests cover the critical paths.

`next_owner=Council` | `stage=council` | Please review and post verdict.

**Comment by markus-lassfolk** [comment]
Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1441
Issue/acceptance: fix: skill write atomicity
Head SHA: 0ac1e4b4771c33152486e0518b40e10b6caa9392
Current state: CI=green, threads=0 resolved, blockers=none, labels=stage/council
Review request: verify current head against acceptance criteria and repo policy.
Output: APPROVED or CHANGES_REQUESTED with exact blockers. Post verdict tied to head SHA.

**Comment by markus-lassfolk** [comment]
Promoting current head `c37b66527b525a7e54618a2e18d9c1c66b2da920` to `stage/approved`: stage/council gate passed in no-merge mode with positive Council/deep evidence, no blocker labels, no unresolved threads, and checks accepted by merge-check. Next owner: merge gate.
